tests/test_text_replacement.py

```diff
@@ -0,0 +1,216 @@
+import unittest
+
+from kmk.keys import ALL_ALPHAS, ALL_NUMBERS, KC
+from kmk.modules.text_replacement import Character, Phrase, Rule, TextReplacement
+from tests.keyboard_test import KeyboardTest
+
+
+class TestTextReplacement(unittest.TestCase):
+    def setUp(self) -> None:
+        self.symbols = '`-=[]\\;\',./~!@#$%^&*()_+{}|:\"<>?'
+        self.everything = ALL_NUMBERS + ALL_ALPHAS + ALL_ALPHAS.lower() + self.symbols
+        self.test_dictionary = {'aa': 'b', 'b': 'aa', '!': '@', 'dccc': 'dcbb'}
+        self.text_replacement = TextReplacement(self.test_dictionary)
+        self.keyboard = KeyboardTest(
+            [self.text_replacement],
+            [
+                [KC.A, KC.B, KC.N1, KC.LSHIFT, KC.LCTRL, KC.C, KC.D],
+            ],
+            debug_enabled=True,
+        )
+
+        return super().setUp()
+
+    def test_keyboard_events_are_correct(self):
```

Comment:
I've implemented all the tests you mentioned, but a particularly long test to test multiple backspaces only passes when I spam `self.keyboard._main_loop()` at the end of `def do_main_loop(self):` in `keyboard_test.py`, as mentioned in #506. Since the test does pass when I do that, I could comment that particular test out and push my changes - how would you like me to proceed?

Reply:
Forgot to mention: Every test that relies on the keyboard doing things after the last key event (timeouts, etc...) need a short delay after the last key event:
```python
        self.keyboard.test(
            'backspace is only tapped as many times as necessary to delete the difference between the key and value',
￼            [ ...
￼                (5, True),
￼                (5, False),
                50,
￼            ], ...
)
```